Sources/FoundationEssentials/Data/Data+Base64.swift

```diff
@@ -0,0 +1,374 @@
+\u{7f}\u{1a}\u{1b}\u{1c}\u{1d}\u{1e}\u{1f}\u{20}\u{21}\u{22}\u{23}\u{24}\u{25}\u{26}\u{27}\u{28}\
+\u{29}\u{2a}\u{2b}\u{2c}\u{2d}\u{2e}\u{2f}\u{30}\u{31}\u{32}\u{33}\u{7f}\u{7f}\u{7f}\u{7f}\u{7f}
+"""
+        assert(base64Decode.isASCII)
+        assert(base64Decode.utf8CodeUnitCount == 128)
+        assert(base64Decode.hasPointerRepresentation)
+
+        let ignoreUnknown = options.contains(.ignoreUnknownCharacters)
+
+        var currentByte: UInt8 = 0
+        var validCharacterCount = 0
+        var paddingCount = 0
+        var index = 0
+
+        for base64Char in bytes {
+            var value: UInt8 = 0
+
+            var invalid = false
+            if base64Char >= base64Decode.utf8CodeUnitCount {
+                invalid = true
+            } else {
+                value = base64Decode.utf8Start[Int(base64Char)]
+                if value & 0x40 == 0x40 {       // Input byte is either '=' or an invalid value.
+                    if value == 0x7f {
+                        invalid = true
+                    } else if value == 0x40 {   // '=' padding at end of input.
+                        paddingCount += 1
+                        continue
+                    }
+                }
+            }
+
+            if invalid {
+                if ignoreUnknown {
+                    continue
+                } else {
+                    throw Base64Error.cannotDecode
+                }
+            }
+            validCharacterCount += 1
+
+            // Padding found in the middle of the sequence is invalid.
+            if paddingCount > 0 {
+                throw Base64Error.cannotDecode
+            }
+
+            switch index {
+            case 0:
+                currentByte = (value << 2)
+            case 1:
+                currentByte |= (value >> 4)
+                output.appendElement(currentByte)
+                currentByte = (value << 4)
+            case 2:
+                currentByte |= (value >> 2)
+                output.appendElement(currentByte)
+                currentByte = (value << 6)
+            case 3:
+                currentByte |= value
+                output.appendElement(currentByte)
+                index = -1
+            default:
+                fatalError("Invalid state")
+            }
+
+            index += 1
+        }
+
+        guard (validCharacterCount + paddingCount) % 4 == 0 else {
+            // Invalid character count of valid input characters.
+            throw Base64Error.cannotDecode
+        }
+    }
+
+    /**
+     This method encodes data in Base64.
+
+     - parameter dataBuffer: A BufferView of the bytes to encode
+     - parameter options:    Options for formatting the result
+     - parameter buffer:     The buffer to write the bytes into
+
+       NOTE: dataBuffer would be better expressed as a <T: Collection> where T.Element == UInt8, T.Index == Int but this currently gives much poorer performance.
```

Comment:
Interesting - how much worse is the performance? Any ideas about what's going on?

The way this function reads data from the buffer is probably not optimal and could allow dropping the `T.Index == Int` requirement.